azure-iot-hub/azure/iot/hub/iothub_amqp_client.py

```diff
@@ -0,0 +1,72 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import os
+import sys
+import base64
+import time
+import hashlib
+import hmac
+from uuid import uuid4
+import urllib
+
+try:
+    from urllib import quote, quote_plus, urlencode  # Py2
+except Exception:
+    from urllib.parse import quote, quote_plus, urlencode
+
+import uamqp
+from uamqp import utils, errors
+
+default_sas_expiry = 30
+
+
+def _generate_auth_token(uri, sas_name, sas_value):
+    """
+    Given a URI, a sas_name, and a sas_value, return a shared access signature.
+    """
+    sas = base64.b64decode(sas_value)
+    expiry = str(int(time.time() + default_sas_expiry))
+    string_to_sign = (uri + "\n" + expiry).encode("utf-8")
+    signed_hmac_sha256 = hmac.HMAC(sas, string_to_sign, hashlib.sha256)
+    signature = urllib.parse.quote(base64.b64encode(signed_hmac_sha256.digest()))
+    return "SharedAccessSignature sr={}&sig={}&se={}&skn={}".format(
+        uri, signature, expiry, sas_name
+    )
+
+
+def _build_amqp_endpoint(hostname, shared_access_key_name, shared_access_key):
```

Comment:
>_build_amqp_endpoint [](start = 4, length = 20)

doc

Reply:
Moved these to be methods in the class.

---
In reply to: [379600948](https://github.com/Azure/azure-iot-sdk-python/pull/465#discussion_r379600948) [](ancestors = 379600948)